File: challenges/InterpreterAndCompilers/WhileLanguage/OCaml/while.ml
open Printf
open List

type binOp = 
    | Add | Sub | Mul | Div

type relOp = 
    | LessThan | GreatherThan

type boolOp = 
    | And | Or

type token = 
    | BoolLit of bool | Var of string | Num of int | BinOp of binOp | RelOp of relOp | BoolOp of boolOp | OpenBracket | CloseBracket

type binExp = 
    | Const of int
    | Variable of string
    | BinExp of binExp * binOp * binExp

type boolExp =
    | True
    | False
    | RelExp of binExp * relOp * binExp
    | Compound of boolExp * boolOp * boolExp

type construct =
    | Assignment of string * binExp
    | While of boolExp * construct list
    | If of boolExp * construct list * construct list

module MyVariables = Map.Make(String) 

type variable = string MyVariables.t

let rec take_while p lst = match lst with 
   [] -> []
  | x::xs -> if p x then x :: (take_while p xs) else []

let rec skip_while p lst = match lst with 
   [] -> []
  | x::xs -> if p x then skip_while p xs else x::xs

let span pred xs =
    let rec loop ys acc =
        match ys with
         [] -> (List.rev acc, ys)
        | y :: yt when pred y -> loop yt (y::acc)
        | _ -> (List.rev acc, ys)
    in loop xs []

let splitWhen pred xs =
    let rec dobreak ys acc =
        match ys with
         [] -> (List.rev acc, [])
        | y :: yt when not (pred y) -> dobreak yt (y::acc)
        | _ -> (List.rev acc, ys)
    in dobreak xs []

let rec startsWith xs ys =
    if xs = [] then true
    else if List.hd xs = List.hd ys then startsWith (List.tl xs) (List.tl ys)
    else false

let rec drop n xs =
    match n with
    0 -> xs
    | _ -> drop (n-1) (List.tl xs)

let explode s = 
    let rec loop acc l =
        match String.length s = l with
         true -> List.rev acc
        | false -> loop (s.[l]::acc) (l+1)
    in loop [] 0

let implode l =
  let result = Bytes.create (List.length l) in
  let rec imp i = function
  | [] -> result
  | c :: l -> result.[i] <- c; imp (i + 1) l in
  imp 0 l;;

let rec trimLeadingSpaces = function
    | ' ' :: chs -> trimLeadingSpaces chs
    | chs -> chs
    
let is_alpha = function 'a' .. 'z' -> true | _ -> false
let is_digit = function '0' .. '9' -> true | _ -> false

let rec tokenize xs acc =
    match xs with
      [] -> List.rev acc
    | ' ' :: xt -> tokenize xt acc

    | '(' :: xt -> tokenize xt (OpenBracket :: acc)
    | ')' :: xt -> tokenize xt (CloseBracket :: acc)

    | '+' :: xt -> tokenize xt (BinOp Add :: acc)
    | '-' :: xt -> tokenize xt (BinOp Sub :: acc)
    | '*' :: xt -> tokenize xt (BinOp Mul :: acc)
    | '/' :: xt -> tokenize xt (BinOp Div :: acc)
    | '<' :: xt -> tokenize xt (RelOp LessThan :: acc)
    | '>' :: xt -> tokenize xt (RelOp GreatherThan :: acc)
    | x :: xt  when is_digit x ->
        let n = take_while is_digit xs |> implode |> int_of_string
        and nds = skip_while is_digit xs 
        in tokenize nds (Num n :: acc)
    | x :: xt  when is_alpha x ->
        let (ds, nds) = span is_alpha xs
        in let var = implode ds
        in match var with
         "true" -> tokenize nds (BoolLit true :: acc)
        | "false" -> tokenize nds (BoolLit false :: acc)
        | "and" ->  tokenize nds (BoolOp And :: acc)
        | "or" ->  tokenize nds (BoolOp Or :: acc)
        | _ -> tokenize nds (Var var :: acc)
    | _ -> raise (Failure "Unexpected error in tokenize!")


let rec getEnclosedtokens ts cnt acc =
    match ts, cnt with
     ([], _) -> (List.rev acc, [])
    | (CloseBracket :: ts, 0) -> (List.rev acc, ts)
    | (CloseBracket :: ts, c) -> getEnclosedtokens (CloseBracket::ts) (c-1) acc 
    | (OpenBracket::ts, c) -> getEnclosedtokens (OpenBracket::ts) (c+1) acc 
    | (t::tt, c) -> getEnclosedtokens tt c (t::acc)

let rec createbinExp xs = 
    if startsWith [Num n] xs then 
        let Num n = List.hd xs 
        in Const n
    else if xs = [Var v] then 
        let Var v = List.hd xs
        in Variable v
    else if startsWith [OpenBracket] xs then
        let (enc, rest) = getEnclosedtokens (List. tl xs) 0 []
        in let bexp = createbinExp enc in
    
        match rest with
         [] -> bexp
        | BinOp op1::rest' -> BinExp (bexp, op1, (createbinExp rest'))
        | _ -> raise (Failure "Error in createbinExp!")

    else if startsWith [Num l :: BinOp Mul :: Num r] xs then
        createbinExp (Num (l*r) :: (drop 3 xs))
    else if startsWith [Num l :: BinOp Div :: Num r] xs then 
        createbinExp (Num (l / r) :: (drop 3 xs))
    else if startsWith [Num l :: BinOp op :: OpenBracket] xs then
        let (enc, rest) = getEnclosedtokens (drop 3 xs) 0 []
        in let bexp = BinExp (Const l, op, createbinExp enc) in
       
        match rest with
         [] -> bexp
        | BinOp op1::rest' -> BinExp (bexp, op1, (createbinExp rest'))
        | _ -> raise (Failure "Error in createbinExp!")

    else if startsWith [Var v :: BinOp op :: OpenBracket] xs then
        let (enc, rest) = getEnclosedtokens (drop 3 xs) 0 []
        and brexp = createbinExp enc in
   
        match rest with
         [] -> BinExp (Variable v, op, brexp)
        | BinOp op1::rest' ->
            if op1 = Mul || op1 = Div
                then BinExp (Variable v, op, BinExp (brexp, op1, createbinExp rest')) 
                else BinExp (BinExp (Variable v, op, brexp), op1, createbinExp rest')        
        | _ -> raise (Failure "Error in createbinExp!")

    else if startsWith [Num l :: BinOp op :: Num r] xs then
        BinExp (Const l, op, (createbinExp (Num r :: (drop 3 xs))))
    else if startsWith [Var lv :: BinOp op :: Var rv] xs then
        BinExp (Variable lv, op, Variable rv)
    else if startsWith [Var lv :: BinOp Mul :: Var rv :: BinOp op2] xs then
        BinExp (BinExp (Variable lv, Mul, Variable rv), op2, createbinExp (drop 4 xs))
    else if startsWith [Var lv :: BinOp Div :: Var rv :: BinOp op2] xs then
        BinExp (BinExp (Variable lv, Div, Variable rv), op2, createbinExp (drop 4 xs))
    else if startsWith [Var lv :: BinOp Mul :: Num r :: BinOp op2] xs then
        BinExp (BinExp (Variable lv, Mul, Const r), op2, createbinExp (drop 4 xs))
    else if startsWith [Var lv :: BinOp Div :: Num r :: BinOp op2] xs then
        BinExp (BinExp (Variable lv, Div, Const r), op2, createbinExp (drop 4 xs))
    else if startsWith [Var v :: BinOp op :: Num r] xs then
        BinExp (Variable v, op, createbinExp (Num r :: (drop 3 xs)))
    else if startsWith [Num l :: BinOp op :: Var v] xs then
        BinExp (Const l, op, createbinExp (Var v :: (drop 3 xs)))
    else if startsWith [Var lv :: BinOp op] xs then
        BinExp (Variable lv, op, createbinExp (drop 2 xs))
    else
        raise (Failure "Unexpected error in createbinExp!")


let findVar (v:string) (map:variable) =
    MyVariables.find v map |> int_of_string

let rec evalBinExp (map:variable) exp =
    match exp with 
     Const l -> l
    | Variable v -> findVar v map
    | BinExp (Const l, Add, Const r) -> l + r
    | BinExp (Const l, Sub, Const r) -> l - r
    | BinExp (Const l, Mul, Const r) -> l * r
    | BinExp (Const l, Div, Const r) -> l / r
    | BinExp (Variable v, op, Const r) -> evalBinExp map (BinExp (Const (findVar v map), op, Const r))
    | BinExp (Variable vl, op, Variable vr) -> evalBinExp map (BinExp (Const (findVar vl map), op, Const (findVar vr map)))
    | BinExp (Const l, op, BinExp(bexp1,bop,bexp2)) -> 
        let r = evalBinExp map (BinExp(bexp1,bop,bexp2))
        in evalBinExp map (BinExp (Const l, op, Const r))
    | BinExp (Variable v, op, BinExp(bexp1,bop,bexp2)) -> 
        let l = findVar v map
        and r = evalBinExp map (BinExp(bexp1,bop,bexp2)) in
        evalBinExp map (BinExp (Const l, op, Const r))
    | BinExp (BinExp(bexp1,bop,bexp2), op, Const r) -> 
        let l = evalBinExp map (BinExp(bexp1,bop,bexp2))
        in evalBinExp map (BinExp (Const l, op, Const r))
    | BinExp (BinExp(bexp1,bop,bexp2), op, Variable v) -> 
        let l = evalBinExp map (BinExp(bexp1,bop,bexp2))
        and  r = findVar v map
        in evalBinExp map (BinExp (Const l, op, Const r))
    | BinExp (BinExp(bexp1,bop,bexp2), op, BinExp(bexp21,bop2,bexp22)) -> 
        let l = evalBinExp map (BinExp(bexp1,bop,bexp2))
        and r = evalBinExp map (BinExp(bexp21,bop2,bexp22))
        in evalBinExp map (BinExp (Const l, op, Const r))
    | _ -> raise (Failure "Unexpected error in evalbinExp!")

let createRelExp (ts:token list) = 
    let xs = take_while (fun t -> t <> RelOp LessThan && t <> RelOp GreatherThan) ts
    and ys = skip_while (fun t -> t <> RelOp LessThan && t <> RelOp GreatherThan) ts

    in match ys with
    | [] -> True
    | RelOp relop :: ys -> RelExp (createbinExp xs, relop, createbinExp ys)
    | _ -> raise (Failure "Unexpected error in createRelExp!")

let rec createboolExp (ts:token list) = 
    let (xs, ys) = splitWhen (fun t -> t = BoolOp And || t = BoolOp Or) ts
    
    in 
    match ys with
     [] -> createRelExp xs 
    | y :: yt ->
        let bexp1 = createRelExp xs
        and bexp2 = createboolExp yt
        and boolop = if y = BoolOp And then And else Or
        in Compound (bexp1, boolop, bexp2)

let rec evalBoolExp map boolexp =
    match boolexp with 
     True -> true
    | False -> false
    | RelExp (bexp1, LessThan, bexp2) -> evalBinExp map bexp1 < evalBinExp map bexp2
    | RelExp (bexp1, GreatherThan, bexp2) -> evalBinExp map bexp1 > evalBinExp map bexp2
    | Compound (bexp1, And, bexp2) -> evalBoolExp map bexp1 && evalBoolExp map bexp2
    | Compound (bexp1, Or, bexp2) -> evalBoolExp map bexp1 || evalBoolExp map bexp2

let rec readBlock cnt (s:char list) (acc:char list) =
    match (cnt, s) with
     (_, []) -> (List.rev acc, [])
    | (_, '{'::ls) -> readBlock (cnt+1) ls ('{'::acc)
    | (1, ('}'::ls)) -> (List.rev acc, ls)
    | (_, ('}'::ls)) -> readBlock (cnt-1) ls ('}'::acc)
    | (_, (ch::ls)) -> readBlock cnt ls (ch::acc)

let rec readUntil s acc =
    match s with 
    [] -> acc, []
    | (')'::ls) -> acc |> trimLeadingSpaces |> List.rev |> trimLeadingSpaces, ls
    | ch :: ls -> readUntil ls (ch::acc)


let rec parseCode (cs: construct list) (code:char list) = 

    let rec getStatements (chs:char list) (acc:char list) (cons: construct list) =
        match (chs, acc) with
         ([], []) -> List.rev cons
        | ([], acc) -> 
            let cs = parseCode [] (acc |> trimLeadingSpaces |> List.rev) 
            in List.rev (List.concat [cs; cons])

        | (';'::xs, acc) -> 
            let cs = parseCode [] (acc |> trimLeadingSpaces |> List.rev) 
            in getStatements xs [] (List.concat [cs; cons])
        
        | (' '::xs, []) -> getStatements xs [] cons
        | (ch :: xs, acc) -> getStatements xs (ch::acc) cons

    in let readStatementsBlock xs =
        let (ss, xs') = readBlock 1 xs []
        in (parseCode [] ss, xs')

    in let rec getAssignment (code:char list) (var:char list) =
    
        match code with
         ':'::'='::xs -> (var |> trimLeadingSpaces |> List.rev |> trimLeadingSpaces, xs)
        | ch :: xs -> getAssignment xs (ch::var)
        | _ -> raise (Failure "Unexpected Assignment!") in

    if List.length code = 0 then 
        List.rev cs
    else if List.hd code = ';' then 
        parseCode cs (List.tl code)
    else if startsWith ['w'; 'h'; 'i'; 'l'; 'e'] code then
        let xt0 = drop 5 code |> trimLeadingSpaces
        in let (bexp, xt1) = readUntil xt0 []
        in match trimLeadingSpaces xt1 with
         ('d'::('o'::xt2)) -> 
            match trimLeadingSpaces xt2 with
             '{'::xt3 -> 
                let (ss, xt4) = readStatementsBlock (trimLeadingSpaces xt3)
                and ts = tokenize bexp []
                in parseCode (While(createboolExp ts, ss) :: cs) xt4
            | _ -> []
        | _ -> []

    else if startsWith [ 'i'; 'f'] code then
        let (_::xt0) = drop 2 code |> trimLeadingSpaces
        in let (bexp, xt1) = readUntil xt0 []
        in match trimLeadingSpaces xt1 with
         't'::'h'::'e'::'n'::xt2 ->
            match trimLeadingSpaces xt2 with
             '{'::xt3 -> 
                let (ifss, xt4) = readStatementsBlock (trimLeadingSpaces xt3)
                in match trimLeadingSpaces xt4 with
                 'e'::'l'::'s'::'e'::xt5 ->
                    match trimLeadingSpaces xt5 with
                     '{'::xt6 -> 
                        let (elsess, xt7) = readStatementsBlock (trimLeadingSpaces xt6)
                        in parseCode (If (createboolExp (tokenize bexp []), ifss, elsess)::cs) xt7   
                    | _ -> raise (Failure "Error in parse code!")
                | _ -> raise (Failure "Error in parse code!")
            | _ -> raise (Failure "Error in parse code!")
        | _ -> raise (Failure "Error in parse code!")
       
    else
        let (xs, code') = splitWhen (fun ch -> ch = ';') code
        in let (var, expr) = getAssignment (trimLeadingSpaces xs) []
        in let ex = createbinExp (tokenize expr [])
        in parseCode (Assignment(implode var, ex) :: cs) code' 
        
let rec evaluateConstructs map cs = 
    match cs with
     [] -> map
    | Assignment (var, bexp) :: cs' ->
        let ev = evalBinExp map bexp
        in evaluateConstructs (MyVariables.add var (string_of_int ev) map) cs'
    | While (bexp, wcs) :: cs' ->
        if evalBoolExp map bexp then 
            evaluateConstructs (evaluateConstructs map wcs) (While (bexp, wcs) :: cs')
        else 
            evaluateConstructs map cs'
    | If (bexp, ifcs, elsecs) :: cs ->
        evaluateConstructs 
            (if evalBoolExp map bexp 
                then evaluateConstructs map ifcs
                else evaluateConstructs map elsecs) cs    



let maybe_read_line () =
  try Some(read_line())
  with End_of_file -> None

let rec loop acc =
  match maybe_read_line () with
   None -> List.rev acc |> String.concat "" 
  | Some(line) -> loop (line :: acc)

let printKeyValue k v =
    print_string k;
    print_string " ";
    print_string v
    

let () = 
    loop []
    |> explode
    |> parseCode []
    |> evaluateConstructs (MyVariables.empty)
    |> MyVariables.iter printKeyValue

File: challenges/InterpreterAndCompilers/WhileLanguage/Scala/While/src/main/scala/While.scala
import scala.annotation.tailrec
import scala.collection.immutable.ListMap
import scala.io.StdIn

sealed abstract class BinOp

case class AddOp() extends BinOp

case class SubOp() extends BinOp

case class MulOp() extends BinOp

case class DivOp() extends BinOp

sealed abstract class RelOp

case class LessThanOp() extends RelOp

case class GreaterThanOp() extends RelOp

sealed abstract class BoolOp

case class AndOp() extends BoolOp

case class OrOp() extends BoolOp

sealed abstract class Token

case class Var(s: String) extends Token

case class Num(n: BigInt) extends Token

case class OpenBracket() extends Token

case class CloseBracket() extends Token

case class True() extends Token

case class False() extends Token

case class And() extends Token

case class Or() extends Token

case class GreaterThan() extends Token

case class LessThan() extends Token

case class Add() extends Token

case class Sub() extends Token

case class Mul() extends Token

case class Div() extends Token

sealed abstract class BinExp

case class Const(n: BigInt) extends BinExp

case class Variable(v: String) extends BinExp

case class Exp(exp1: BinExp, op: BinOp, exp2: BinExp) extends BinExp

sealed abstract class BoolExp

case class TrueExp() extends BoolExp

case class FalseExp() extends BoolExp

case class RelExp(exp1: BinExp, op: RelOp, exp2: BinExp) extends BoolExp

case class CompoundExp(exp1: BoolExp, op: BoolOp, exp2: BoolExp) extends BoolExp

sealed abstract class Construct

case class Assignment(v: String, exp: BinExp) extends Construct

case class While(cond: BoolExp, constructs: Seq[Construct]) extends Construct

case class If(cond: BoolExp, ifSeq: Seq[Construct], elseSeq: Seq[Construct]) extends Construct

object While extends App {

  def tokenize(xs: Seq[Char]) = {

    def parseChars(s: Seq[Char]) = {

      def handleAlphabets(s1: Seq[Char]) = {
        s1.takeWhile(Character.isLetter).filterNot(_.isWhitespace) match {
          case Seq('t', 'r', 'u', 'e') => (True(), 4)
          case Seq('f', 'a', 'l', 's', 'e') => (False(), 5)
          case Seq('a', 'n', 'd') => (And(), 3)
          case Seq('o', 'r') => (Or(), 2)
          case v => (Var(v.filterNot(_.isWhitespace).mkString), v.length)
        }
      }

      val s1 = s.dropWhile(_ == ' ')

      if (s1.isEmpty) {
        None
      } else {
        val (token, dropCount) = s1.head match {
          case '(' => (OpenBracket(), 1)
          case ')' => (CloseBracket(), 1)

          case '+' => (Add(), 1)
          case '-' => (Sub(), 1)
          case '*' => (Mul(), 1)
          case '/' => (Div(), 1)

          case '<' => (LessThan(), 1)
          case '>' => (GreaterThan(), 1)

          case _ =>
            if (Character.isDigit(s1.head)) {
              val pre = s.takeWhile(Character.isDigit)
              val num = BigInt(pre.mkString)
              (Num(num), pre.length)
            }
            else if (Character.isLetter(s1.head)) {
              handleAlphabets(s1)
            }
            else {
              throw new Exception("Unexpected character!")
            }
        }
        Some(token, s1.drop(dropCount).dropWhile(c => c == ' '))
      }
    }

    Seq.unfold(xs.dropWhile(_ == ' '))(parseChars)

  }

  def getAssignment(value: Seq[Char]): (Seq[Char], Seq[Char]) = {
    (Seq.empty, Seq.empty)
  }

  @tailrec
  def getEnclosedTokens(ts: Seq[Token], cnt: Int, acc: Seq[Token]): (Seq[Token], Seq[Token]) = {
    if (ts.isEmpty)
      (acc, Seq.empty)
    else
      ts.head match {
        case CloseBracket() if cnt == 0 => (acc, ts.tail)
        case CloseBracket() => getEnclosedTokens(ts, cnt - 1, acc)
        case OpenBracket() => getEnclosedTokens(ts, cnt + 1, acc)
        case _ => getEnclosedTokens(ts.tail, cnt, acc :+ ts.head)
      }
  }

  def createBinExp(ts: Seq[Token]): BinExp = {
    def getOp(token: Token) = {
      token match {
        case Add() => AddOp()
        case Sub() => SubOp()
        case Mul() => MulOp()
        case Div() => DivOp()
      }
    }

    ts match {
      case Seq(Num(n)) => Const(n)
      case Seq(Var(v)) => Variable(v)
      case Seq(OpenBracket(), _*) =>
        val (enc, rest) = getEnclosedTokens(ts.tail, 0, Seq.empty)
        val binExp = createBinExp(enc)
        if (rest.isEmpty)
          binExp
        else
          rest.head match {
            case Add() | Sub() | Mul() | Div() => Exp(binExp, getOp(rest.head), createBinExp(rest.tail))
            case _ => throw new Exception("Error while creating binary expression")
          }
      case Seq(Num(l), Mul(), Num(r), _*) => createBinExp(ts.drop(3) :+ Num(l * r))
      case Seq(Num(l), Div(), Num(r), _*) => createBinExp(ts.drop(3) :+ Num(l / r))

      case Seq(Num(l), op@(Add() | Sub() | Mul() | Div()), OpenBracket(), _*) =>
        val (enc, rest) = getEnclosedTokens(ts.drop(3), 0, Seq.empty)
        val binExp = Exp(Const(l), getOp(op), createBinExp(enc))
        rest match {
          case _ if rest.isEmpty => binExp
          case Seq(op1@(Add() | Sub() | Mul() | Div()), _*) => Exp(binExp, getOp(op1), createBinExp(rest.tail))
        }

      case Seq(Var(v), op@(Add() | Sub() | Mul() | Div()), OpenBracket(), _*) =>
        val (enc, rest) = getEnclosedTokens(ts.drop(3), 0, Seq.empty)
        val binExp = Exp(Variable(v), getOp(op), createBinExp(enc))
        rest match {
          case _ if rest.isEmpty => binExp
          case Seq(op1@(Mul() | Div()), _*) => Exp(Variable(v), getOp(op), Exp(createBinExp(enc), getOp(op1), createBinExp(rest.tail)))
          case Seq(op1@(Add() | Sub() | Mul() | Div()), _*) => Exp(Exp(Variable(v), getOp(op), createBinExp(enc)), getOp(op1), createBinExp(rest.tail))
        }

      case Seq(Num(l), op@(Add() | Sub() | Mul() | Div()), Num(r), _*) =>
        Exp(Const(l), getOp(op), createBinExp(Num(r) +: ts.drop(2)))

      case Seq(Var(lv), op@(Add() | Sub() | Mul() | Div()), Var(rv)) =>
        Exp(Variable(lv), getOp(op), Variable(rv))

      case Seq(Var(lv), op@(Mul() | Div()), Var(rv), op2@(Add() | Sub() | Mul() | Div()), _*) =>
        Exp(Exp(Variable(lv), getOp(op), Variable(rv)), getOp(op2), createBinExp(ts.drop(4)))

      case Seq(Var(lv), op@(Mul() | Div()), Num(r), op2@(Add() | Sub() | Mul() | Div()), _*) =>
        Exp(Exp(Variable(lv), getOp(op), Const(r)), getOp(op2), createBinExp(ts.drop(4)))

      case Seq(Var(lv), op@(Add() | Sub() | Mul() | Div()), Num(r), _*) =>
        Exp(Variable(lv), getOp(op), createBinExp(Num(r) +: ts.drop(3)))

      case Seq(Num(l), op@(Add() | Sub() | Mul() | Div()), Var(rv), _*) =>
        Exp(Const(l), getOp(op), createBinExp(Var(rv) +: ts.drop(3)))

      case Seq(Var(v), op@(Add() | Sub() | Mul() | Div()), _*) =>
        Exp(Variable(v), getOp(op), createBinExp(ts.drop(2)))

    }
  }

  def evalBinExp(map: Map[String, BigInt], exp: BinExp): BigInt = {
    exp match {
      case Const(l) => l
      case Variable(v) => map(v.filterNot(_.isWhitespace))
      case Exp(Const(l), AddOp(), Const(r)) => l + r
      case Exp(Const(l), SubOp(), Const(r)) => l - r
      case Exp(Const(l), MulOp(), Const(r)) => l * r
      case Exp(Const(l), DivOp(), Const(r)) => l / r
      case Exp(Variable(v), op, Const(r)) => evalBinExp(map, Exp(Const(map(v)), op, Const(r)))
      case Exp(Variable(lv), op, Variable(rv)) => evalBinExp(map, Exp(Const(map(lv)), op, Const(map(rv))))
      case Exp(Const(c), op, exp@Exp(_, _, _)) =>
        val r = evalBinExp(map, exp)
        evalBinExp(map, Exp(Const(c), op, Const(r)))
      case Exp(Variable(v), op, exp@Exp(_, _, _)) =>
        val l = map(v)
        val r = evalBinExp(map, exp)
        evalBinExp(map, Exp(Const(l), op, Const(r)))
      case Exp(exp@Exp(_, _, _), op, Const(c)) =>
        val r = evalBinExp(map, exp)
        evalBinExp(map, Exp(Const(c), op, Const(r)))
      case Exp(exp@Exp(_, _, _), op, Variable(v)) =>
        val l = map(v)
        val r = evalBinExp(map, exp)
        evalBinExp(map, Exp(Const(l), op, Const(r)))
      case Exp(exp1@Exp(_, _, _), op, exp2@Exp(_, _, _)) =>
        val l = evalBinExp(map, exp1)
        val r = evalBinExp(map, exp2)
        evalBinExp(map, Exp(Const(l), op, Const(r)))
    }
  }

  def createRelExp(ts: Seq[Token]): BoolExp = {
    val xs = ts.takeWhile(t => t != LessThan() && t != GreaterThan())
    val ys = ts.drop(xs.length).dropWhile(_ == ' ')

    if (ys.isEmpty)
      TrueExp()
    else
      ys.head match {
        case LessThan() => RelExp(createBinExp(xs), LessThanOp(), createBinExp(ys.tail.dropWhile(_ == ' ')))
        case GreaterThan() => RelExp(createBinExp(xs), GreaterThanOp(), createBinExp(ys.tail.dropWhile(_ == ' ')))
      }
  }

  def createBoolExp(ts: Seq[Token]): BoolExp = {
    val xs = ts.takeWhile(t => t != And() && t != Or())
    val ys = ts.drop(xs.length)

    if (ys.isEmpty)
      createRelExp(xs)
    else {
      val exp1 = createRelExp(xs)
      val exp2 = createBoolExp(ys.tail)
      val op = ys.head match {
        case And() => AndOp()
        case Or() => OrOp()
      }
      CompoundExp(exp1, op, exp2)
    }
  }

  def evalBoolExp(map: Map[String, BigInt], exp: BoolExp): Boolean = {
    exp match {
      case TrueExp() => true
      case FalseExp() => false
      case RelExp(exp1, LessThanOp(), exp2) => evalBinExp(map, exp1) < evalBinExp(map, exp2)
      case RelExp(exp1, GreaterThanOp(), exp2) => evalBinExp(map, exp1) > evalBinExp(map, exp2)
      case CompoundExp(exp1, AndOp(), exp2) => evalBoolExp(map, exp1) && evalBoolExp(map, exp2)
      case CompoundExp(exp1, OrOp(), exp2) => evalBoolExp(map, exp1) || evalBoolExp(map, exp2)
    }
  }

  def parseCode(code: Seq[Char], cs: Seq[Construct]): Seq[Construct] = {

    @tailrec
    def readBlock(cnt: Int, s: Seq[Char], acc: Seq[Char]): (Seq[Char], Seq[Char]) = {
      if (s.isEmpty && acc.isEmpty) {
        (acc, Seq.empty)
      }
      else {
        (cnt, s.head) match {
          case (1, '}') => (acc, s.tail)
          case (_, '}') => readBlock(cnt - 1, s.tail, acc :+ s.head)
          case (_, '{') => readBlock(cnt + 1, s.tail, acc :+ s.head)
          case (_, _) => readBlock(cnt, s.tail, acc :+ s.head)
        }
      }
    }

    //    def getStatements (chs: Seq[Char]) = {
    //
    //      def readStatement(chs:Seq[Char]) : Option[(Seq[Construct], Seq[Char])] = {
    //        val (pre, suf) = chs.span(c => c != ';')
    //        if (trimLeadingSpaces(pre).isEmpty) {
    //          None
    //        } else {
    //          Some(parseCode(pre, Seq.empty), suf)
    //        }
    //      }
    //
    //      Seq.concat(Seq.unfold(chs)(readStatement))
    //
    //    }

    def readStatementsBlock(xs: Seq[Char]) = {
      val (block, rest) = readBlock(1, xs, Seq.empty)
      (parseCode(block, Seq.empty), rest)
    }

    @tailrec
    def getAssignment(code: Seq[Char], variable: Seq[Char]): (Seq[Char], Seq[Char]) = {
      code match {
        case Seq(':', '=', _*) => (variable.dropWhile(_ == ' '), code.drop(2).filterNot(_.isWhitespace))
        case _ => getAssignment(code.tail.filterNot(_.isWhitespace), if (code.head.isWhitespace) variable else variable :+ code.head)
      }
    }

    def parseWhileConstruct(xs: Seq[Char]) = {
      val (cond, exp) = xs.span(c => c != ')') //skip the end bracket
      val exp1 = exp.tail.dropWhile(c => c == ' ')
      exp1 match {
        case Seq('d', 'o', _*) =>
          val exp2 = exp1.drop(2).dropWhile(c => c == ' ' || c == '{')
          val (ss, exp3) = readStatementsBlock(exp2)
          val boolExp = createBoolExp(tokenize(cond))
          parseCode(exp3, cs :+ While(boolExp, ss))

        case _ => throw new Exception("Parse error!")
      }

    }

    def parseIfConstruct(xs: Seq[Char]) = {
      val (cond, exp) = xs.dropWhile(c => c == '(' || c == ' ').span(c => c != ')')
      val exp1 = exp.dropWhile(c => c == ')' || c == ' ')

      exp1 match {
        case Seq('t', 'h', 'e', 'n', _*) =>
          val (ifSeq, exp3) = readStatementsBlock(exp1.drop(4).dropWhile(c => c == '{' || c == ' '))
          val exp4 = exp3.dropWhile(c => c == ')' || c == ' ')
          exp4 match {
            case Seq('e', 'l', 's', 'e', _*) =>
              val (elseSeq, exp6) = readStatementsBlock(exp4.drop(4).dropWhile(c => c == '{' || c == ' '))
              parseCode(exp6, cs :+ If(createBoolExp(tokenize(cond)), ifSeq, elseSeq))
          }
        case _ => throw new Exception("Error in parsing if statement")
      }
    }


    def parseAssignmentConstruct(xs: Seq[Char]) = {
      val (v, expr) = getAssignment(xs, Seq.empty)
      val ex = createBinExp(tokenize(expr))
      Assignment(v.mkString, ex)
    }

    val chs = code.dropWhile(c => c == ' ' || c == ';' || c == '}')
    chs match {
      case _ if chs.isEmpty => cs
      case Seq('w', 'h', 'i', 'l', 'e', _*) => parseWhileConstruct(chs.drop(5).dropWhile(c => c == ' ' || c == '('))
      case Seq('i', 'f', _*) => parseIfConstruct(chs.drop(2).dropWhile(c => c == ' ' || c == '('))
      case xs =>
        val (pre, suf) = xs.span(c => c != ';')
        val ac = parseAssignmentConstruct(pre)
        parseCode(suf, cs :+ ac)
    }

  }

  def evaluateConstructs(map: Map[String, BigInt], cs: Seq[Construct]): Map[String, BigInt] = {
    if (cs.isEmpty)
      map
    else
      cs.head match {
        case Assignment(v, exp) =>
          evaluateConstructs(map  + (v -> evalBinExp(map, exp)), cs.tail)
        case While(cond, wcs) =>
          if (evalBoolExp(map, cond))
            evaluateConstructs(evaluateConstructs(map, wcs), cs)
          else
            evaluateConstructs(map, cs.tail)
        case If(cond, ifSeq, elseSeq) =>
          evaluateConstructs(
            if (evalBoolExp(map, cond))
              evaluateConstructs(map, ifSeq)
            else
              evaluateConstructs(map, elseSeq), cs.tail)
      }
  }

  def readAllLines() = {

    @tailrec
    def reread(xs: Seq[Char]): Seq[Char] = {

      val s = StdIn.readLine()

      s match {
        case _ if s.startsWith(":exit") => xs
        case null => xs
        case _ => reread(xs ++ s.toSeq)
      }
    }

    reread(Seq.empty)

  }

  val all = readAllLines()

  val cs = parseCode(all, Seq.empty)
  val map = evaluateConstructs(Map.empty, cs)
  ListMap(map.toSeq.sortBy(_._1):_*).foreach(kv => {
    System.out.print(kv._1)
    System.out.print(' ')
    System.out.println(kv._2)
  })

}
